### ISSUE TITLE
GdNavigationServer crash

### ISSUE THREAD

**Issue by tavurth** [godotengine/godot#59218]
### Godot version

v3.5.beta2.official.7a4f9dfb1

### System information

Mac OSX (Intel 2019)

### Issue description

Crash when freeing NavMesh just after starting build.

Seems related to this issue: https://github.com/godotengine/godot/issues/46285

```
~/src/godot-navigation-bug
▶ godot
arguments
0: /Applications/Godot.app/Contents/MacOS/Godot
Current path: /Users/will/src/godot-navigation-bug
Godot Engine v3.5.beta2.official.7a4f9dfb1 - https://godotengine.org
OpenGL ES 2.0 Renderer: Intel(R) Iris(TM) Plus Graphics 655

Registered camera FaceTime HD Camera (Built-in) with id 1 position 0 at index 0
Exiting NavMesh
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
Adding Navigation
Exiting NavMesh
ERROR: Condition "!is_inside_tree()" is true. Returned: Transform()
   at: get_global_transform (scene/3d/spatial.cpp:385)
ERROR: Condition "!is_inside_tree()" is true. Returned: Transform()
   at: get_global_transform (scene/3d/spatial.cpp:385)
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
Adding Navigation
Exiting NavMesh
ERROR: Condition "!is_inside_tree()" is true. Returned: Transform()
   at: get_global_transform (scene/3d/spatial.cpp:385)
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
Adding Navigation
Exiting NavMesh
ERROR: Condition "!is_inside_tree()" is true. Returned: Transform()
   at: get_global_transform (scene/3d/spatial.cpp:385)
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
Adding Navigation
Exiting NavMesh
ERROR: Condition "!is_inside_tree()" is true. Returned: Transform()
   at: get_global_transform (scene/3d/spatial.cpp:385)
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
Adding Navigation
Exiting NavMesh
ERROR: Condition "!is_inside_tree()" is true. Returned: Transform()
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
   at: get_global_transform (scene/3d/spatial.cpp:385)

================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v3.5.beta2.official (7a4f9dfb15a6817135e753b286cd5b767119d08b)
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] 1   libsystem_platform.dylib            0x00007ff807efde2d _sigtramp + 29
[2] 2   ???                                 0x0000000000000000 0x0 + 0
[3] void process_array_thread<ThreadArrayProcessData<NavMap, RvoAgent**> >(void*)
[4] void process_array_thread<ThreadArrayProcessData<NavMap, RvoAgent**> >(void*)
[5] Object* ClassDB::creator<MobileVRInterface>()
[6] NavigationAgent::get_nav_path_index() const
[7] MutexImpl<std::__1::mutex>::try_lock() const
Adding Navigation
Exiting NavMesh
ERROR: Condition "!is_inside_tree()" is true. Returned: Transform()
   at: get_global_transform (scene/3d/spatial.cpp:385)
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
[8] MutexImpl<std::__1::mutex>::try_lock() const
[9] 9   libsystem_pthread.dylib             0x00007ff807ee84f4 _pthread_start + 125
[10] 10  libsystem_pthread.dylib             0x00007ff807ee400f thread_start + 15
-- END OF BACKTRACE --
================================================================
[1]    25792 abort      /Applications/Godot.app/Contents/MacOS/Godot
```

### Steps to reproduce

Run this project

[godot-navigation-bug.zip](https://github.com/godotengine/godot/files/8276478/godot-navigation-bug.zip)

**Comment by tavurth** [comment]
If you wait one second you still get a thread error:

Project with example:
[godot-navigation-bug.zip](https://github.com/godotengine/godot/files/8276559/godot-navigation-bug.zip)


```
▶ godot
arguments
0: /Applications/Godot.app/Contents/MacOS/Godot
Current path: /Users/will/src/godot-navigation-bug
Godot Engine v3.5.beta2.official.7a4f9dfb1 - https://godotengine.org
OpenGL ES 2.0 Renderer: Intel(R) Iris(TM) Plus Graphics 655

Registered camera FaceTime HD Camera (Built-in) with id 1 position 0 at index 0
Adding Navigation
Exiting NavMesh
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
Exiting NavMesh
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
Adding Navigation
Adding Navigation
Exiting NavMesh
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
Adding Navigation
Exiting @NavMesh@2
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
Adding Navigation
Exiting NavMesh
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
Adding Navigation
Exiting @NavMesh@3
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
Adding Navigation
Exiting NavMesh
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
Adding Navigation
Exiting @NavMesh@4
WARNING: A Thread object has been destroyed without wait_to_finish() having been called on it. Please do so to ensure correct cleanup of the thread.
     at: ~Thread (core/os/thread.cpp:123)
```

**Comment by tavurth** [comment]
If you don't `queue_free()` you get this error

<img width="969" alt="Screenshot 2022-03-17 at 03 03 31" src="https://user-images.githubusercontent.com/100964/158711153-d3dd1acd-d42c-4090-a772-11709c63efcc.png">

Updated sample:
[godot-navigation-bug.zip](https://github.com/godotengine/godot/files/8276657/godot-navigation-bug.zip)

**Comment by Calinou** [comment]
cc @Scony

**Comment by Scony** [comment]
Confirmed on manjaro linux.

The issue is due to poor thread handling. I've already seen it in another corner case. I'll draft a fix in upcoming weeks.

**Comment by akien-mga** [comment]
Fixed by #59849.
